### ISSUE TITLE
[Live Component] DoctrineObjectNormalizer is being used to normalize my non-Doctrine object

### ISSUE THREAD

**Issue by benr77** [symfony/ux#452]
Using v2.3 of the Live Component.

I'm using a DTO as a data class with a form in a live component. It's working OK with scalar values in the DTO.

However, when I try and use `Money/Money` object as one of the form data properties, it borks with:

```
ErrorException:
Warning: Undefined array key 0

  at vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:883
  at Doctrine\ORM\Mapping\ClassMetadataInfo->getIdentifierValues(object(Money))
     (vendor/symfony/ux-live-component/src/Normalizer/DoctrineObjectNormalizer.php:44)
  at Symfony\UX\LiveComponent\Normalizer\DoctrineObjectNormalizer->normalize(object(Money), 'json', array('live-component' => true, 'cache_key' => '87beb01efbb89bc8a83bf4e8e91ba0dc', 'circular_reference_limit_counters' => array('0000000000000d700000000000000000' => 1)))
     (vendor/symfony/serializer/Serializer.php:161)
  at Symfony\Component\Serializer\Serializer->normalize(object(Money), 'json', array('live-component' => true, 'cache_key' => '87beb01efbb89bc8a83bf4e8e91ba0dc', 'circular_reference_limit_counters' => array('0000000000000d700000000000000000' => 1)))
     (vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:216)
  at Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer->normalize(object(DefaultPriceData), 'json', array('live-component' => true, 'cache_key' => '87beb01efbb89bc8a83bf4e8e91ba0dc', 'circular_reference_limit_counters' => array('0000000000000d700000000000000000' => 1)))
     (vendor/symfony/serializer/Serializer.php:161)
  at Symfony\Component\Serializer\Serializer->normalize(object(DefaultPriceData), 'json', array('live-component' => true))
     (vendor/symfony/ux-live-component/src/LiveComponentHydrator.php:87)file=vendor/symfony/serializer/Serializer.php&line=161#line161)-> normalize (line 161)
```

Looking at the `DoctrineObjectNormalizer` I see it is given a priority of 100, making it kick in way before my custom Money normalizer which is already registered with the serializer.

Why is the `DoctrineObjectNormalizer` given such a high priority, and why is it trying to normalize my non-Doctrine object?

What would be the best way to get around this?

Thank you.

**Comment by kbond** [comment]
Hey @benr77, strange, the doctrine one should be skipped as the object is not managed. I'll look into this.

I'm also questioning the priority. Currently, I don't think a custom doctrine entity normalizer can be used.

**Comment by kbond** [comment]
#453 changes the priority from 100 to -100. I added a test that I think covers your scenario above (dto object with a normalizer) but couldn't recreate the error you received.

**Comment by benr77** [comment]
My serializer is currently registered with a priority of -700, which IIRC is to ensure it is checked before `ObjectNormalizer` (priority -1000) but can still be overridden easily by client implementations (it's from a reusable bundle).

But `DoctrineObjectNormalizer` should not agree to normalize a non-Doctrine object surely?

My form data class just looks like this:

```php
use Money\Money;

final class DefaultPriceData
{
    public function __construct(
        public string $itemType,
        public Money $price
    ) {
    }
}
```

Thanks

**Comment by benr77** [comment]
On further investigation, it turns out that `DoctrineObjectNormalizer` does consider `Money\Money` to be managed by Doctrine.

https://github.com/symfony/ux-live-component/blob/17a2586f0fe86c4efbcabd39b3a90bcd453a39b2/src/Normalizer/DoctrineObjectNormalizer.php#L85

This line returns `EntityManager` for `Money\Money`. 

It would seem that this is because I have a Doctrine embeddable defined to allow persisting Money objects to the database.

```xml
<embeddable name="Money\Money">
    <field name="amount" type="decimal" precision="0" scale="0" />
    <field name="currency" type="currency"/>
</embeddable>
```

So I'm guessing that the actual issue is simply that the `DoctrineObjectNormalizer` needs to check it's a Doctrine _entity_ rather than just any old object that Doctrine is aware of.

**Comment by kbond** [comment]
Ahhh, now it makes sense! I'll work on a fix.

**Comment by kbond** [comment]
Hmm, still can't reproduce the issue. In #453, I've added a test that demonstrates using an embeddable as an entity property but also a dto property. I've also confirmed that:

https://github.com/symfony/ux/blob/85760ce651d2d7b883e86c273c5e4eb30d3eec3b/src/LiveComponent/src/Normalizer/DoctrineObjectNormalizer.php#L81

Returns `null` for the embeddable (as we would expect). What version of doctrine/orm are you using? I do think #453 does fix your issue (but only because of the priority change). Can you confirm?

**Comment by benr77** [comment]
I'm using doctrine/orm 2.13.1

Changing the priority for `DoctrineObjectNormalizer` to -100 does not fix the issue (this is all I've changed - not used the rest of #453)

If I remove the code that registers the embeddable from my bundle, the Live Component starts working, but I run into an "Invalid Checksum" error instead. Not sure if this is related or not.

Finally, the code that adds the embeddable does so using `DoctrineOrmMappingsPass` in the bundle file:

```php
public function build(ContainerBuilder $container): void
    {
        parent::build($container);

        $modelDir = realpath(__DIR__ . '/Doctrine/Mapping');
        $mappings = [
            $modelDir => 'Money',
        ];

        if (class_exists(DoctrineOrmMappingsPass::class)) {
            $container->addCompilerPass(
                DoctrineOrmMappingsPass::createXmlMappingDriver(
                    $mappings,
                    ['doctrine.orm.entity_manager'],
                    false
                )
            );
        }
    }
```

Not sure if any of this is helpful sorry.

**Comment by kbond** [comment]
Finally solved! Was a tricky one: when embeddables are mapped using annotations, `ManagerRegistry::getManagerForClass()` returns `null` but when mapped using xml, it returns the `EntityManager`! I've added this scenario to the test in #453.

**Comment by benr77** [comment]
You sir, are a legend.

Isn't this a Doctrine bug then?

**Comment by kbond** [comment]
Found a bit of context. I actually solved this exact thing in [zenstruck/foundry](https://github.com/zenstruck/foundry/pull/271) earlier this year!

Refs:
- https://github.com/zenstruck/foundry/issues/270#issuecomment-1106577221
- https://github.com/doctrine/orm/pull/8138
- https://github.com/doctrine/orm/pull/8006
- https://github.com/doctrine/orm/issues/7570

In short, it looks like it is a bug but fixing it causes other problems so they're leaving alone for now. Sounds like it will all be refactored/fixed in ORM 3.0.
